src/allocator/bootstrap.rs

```diff
@@ -0,0 +1,99 @@
+//! A bootstrap allocator based on a statically allocated buffer.
+
+/// A pointer range that can only be compared against.
+mod ptr_range {
+	use core::ops::Range;
+	use core::ptr::NonNull;
+
+	/// A pointer range that can only be compared against.
+	pub struct PtrRange<T> {
+		inner: Range<NonNull<T>>,
+	}
+
+	// SAFETY: We never dereference, but only compare, pointers.
+	unsafe impl<T> Send for PtrRange<T> {}
+	unsafe impl<T> Sync for PtrRange<T> {}
+
+	impl<T> PtrRange<T> {
+		/// Returns `true` if the pointer range contains `ptr`.
+		pub fn contains(&self, ptr: NonNull<T>) -> bool {
+			self.inner.contains(&ptr)
+		}
+	}
+
+	impl<T> From<Range<NonNull<T>>> for PtrRange<T> {
+		fn from(value: Range<NonNull<T>>) -> Self {
+			Self { inner: value }
+		}
+	}
+}
+
+use core::alloc::{AllocError, Allocator, Layout};
+use core::mem::MaybeUninit;
+use core::ops::Range;
+use core::ptr::NonNull;
+
+use hermit_sync::ExclusiveCell;
```

Comment:
We can use [`exclusive_cell::ExclusiveCell`](https://docs.rs/exclusive_cell/0.1.0/exclusive_cell/index.html) directly here. We don't need all the other types from hermit-sync and can reduce dependencies.